js/src/type.ts

```diff
@@ -0,0 +1,567 @@
+        return new Schema(vectors.map((v, i) => new Field('' + i, v.type)));
+    }
+    // @ts-ignore
+    protected _bodyLength: number;
+    // @ts-ignore
+    protected _headerType: MessageHeader;
+    public readonly fields: Field[];
+    public readonly version: MetadataVersion;
+    public readonly metadata?: Map<string, string>;
+    public readonly dictionaries: Map<number, Field<Dictionary>>;
+    constructor(fields: Field[],
+                metadata?: Map<string, string>,
+                version: MetadataVersion = MetadataVersion.V4,
+                dictionaries: Map<number, Field<Dictionary>> = new Map()) {
+        this.fields = fields;
+        this.version = version;
+        this.metadata = metadata;
+        this.dictionaries = dictionaries;
+    }
+    public get bodyLength() { return this._bodyLength; }
+    public get headerType() { return this._headerType; }
+    public select(...fieldNames: string[]): Schema {
+        const namesToKeep = fieldNames.reduce((xs, x) => (xs[x] = true) && xs, Object.create(null));
+        const newDictFields = new Map(), newFields = this.fields.filter((f) => namesToKeep[f.name]);
+        this.dictionaries.forEach((f, dictId) => (namesToKeep[f.name]) && newDictFields.set(dictId, f));
+        return new Schema(newFields, this.metadata, this.version, newDictFields);
+    }
+    public static [Symbol.toStringTag] = ((prototype: Schema) => {
+        prototype._bodyLength = 0;
+        prototype._headerType = MessageHeader.Schema;
+        return 'Schema';
+    })(Schema.prototype);
+}
+
+export class Field<T extends DataType = DataType> {
+    public readonly type: T;
+    public readonly name: string;
+    public readonly nullable: boolean;
+    public readonly metadata?: Map<string, string> | null;
+    constructor(name: string, type: T, nullable = false, metadata?: Map<string, string> | null) {
+        this.name = name;
+        this.type = type;
+        this.nullable = nullable;
+        this.metadata = metadata;
+    }
+    public toString() { return `${this.name}: ${this.type}`; }
+    public get typeId(): T['TType'] { return this.type.TType; }
+    public get [Symbol.toStringTag](): string { return 'Field'; }
+    public get indicies(): T | Int<any> {
+        return DataType.isDictionary(this.type) ? this.type.indicies : this.type;
+    }
+}
+
+export type TimeBitWidth = 32 | 64;
+export type IntBitWidth = 8 | 16 | 32 | 64;
+
+export type NumericType = Int | Float | Date_ | Time | Interval | Timestamp;
+export type FixedSizeType = Int64 |  Uint64 | Decimal | FixedSizeBinary;
+export type PrimitiveType = NumericType | FixedSizeType;
+
+export type FlatListType = Utf8 | Binary; // <-- these types have `offset`, `data`, and `validity` buffers
+export type FlatType = Bool | PrimitiveType | FlatListType; // <-- these types have `data` and `validity` buffers
+export type ListType = List<any> | FixedSizeList<any>; // <-- these types have `offset` and `validity` buffers
```

Comment:
Its not true that `FixedSizeList` has an offset buffer, which is causing some read errors for me. I'm looking into fixing this myself to try to get familiar with the changes.

Reply:
Ah you're right, looks like that's an oversight. I moved some things around after defining these types the first time, so comments are probably out of date.